107/src/main/java/org/ehcache/jsr107/ConfigurationMerger.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.jsr107;
+
+import org.ehcache.config.CacheConfiguration;
+import org.ehcache.config.CacheConfigurationBuilder;
+import org.ehcache.config.loaderwriter.DefaultCacheLoaderWriterConfiguration;
+import org.ehcache.config.xml.XmlConfiguration;
+import org.ehcache.internal.store.heap.service.OnHeapStoreServiceConfiguration;
+import org.ehcache.spi.loaderwriter.CacheLoaderWriter;
+import org.slf4j.Logger;
+
+import javax.cache.configuration.Configuration;
+
+import static org.ehcache.config.CacheConfigurationBuilder.newCacheConfigurationBuilder;
+
+/**
+ * ConfigurationMerger
+ */
+class ConfigurationMerger {
+
+  private final XmlConfiguration xmlConfiguration;
+  private final Jsr107Service jsr107Service;
+  private final Eh107CacheLoaderWriterFactory cacheLoaderWriterFactory;
+  private final Logger log;
+
+  ConfigurationMerger(org.ehcache.config.Configuration ehConfig, Jsr107Service jsr107Service, Eh107CacheLoaderWriterFactory cacheLoaderWriterFactory, Logger log) {
+    if (ehConfig instanceof XmlConfiguration) {
+      xmlConfiguration = (XmlConfiguration) ehConfig;
+    } else {
+      xmlConfiguration = null;
+    }
+    this.jsr107Service = jsr107Service;
+    this.cacheLoaderWriterFactory = cacheLoaderWriterFactory;
+    this.log = log;
+  }
+
+  <K, V> ConfigHolder<K, V> mergeConfigurations(String cacheName, Configuration<K, V> configuration) {
+    ConfigHolder<K, V> result = new ConfigHolder<K, V>();
+    CacheConfigurationBuilder<K, V> builder = newCacheConfigurationBuilder();
+    boolean useEhcacheExpiry = true;
+    boolean useEhcacheLoaderWriter = true;
+
+    result.jsr107Configuration = new Eh107CompleteConfiguration<K, V>(configuration);
+
+    try {
+      result.cacheResources = new CacheResources<K, V>(cacheName, result.jsr107Configuration);
+
+      String templateName = getCacheTemplateName(cacheName);
+      if (xmlConfiguration != null && templateName != null) {
+        CacheConfigurationBuilder<K, V> templateBuilder = xmlConfiguration.newCacheConfigurationBuilderFromTemplate(templateName,
+                        result.jsr107Configuration.getKeyType(), result.jsr107Configuration.getValueType());
+        if (templateBuilder != null) {
+          builder = templateBuilder;
+          log.debug("Configuration of cache {} will be supplemented by template {}", cacheName, templateName);
```

Comment:
Supplemented doesn't seem like quite the right word any more.  Maybe "combined" or "merged"?